### DISCUSSION TITLE
Android/iOS App to always run in background (like a service)

### DISCUSSION THREAD

**Discussion by 2jfw** [airsdk/Adobe-Runtime-Support#2915] [Q&A]
Referring to this question on Stackoverflow (https://stackoverflow.com/questions/34573109/how-to-make-an-android-app-to-always-run-in-background): 
is it possible to run an AIR app in the background on Android (and also iOS)?

In the thread it is stated:
>You have to start a service in your Application class to run it always. If you do that, your service will be always running. Even though user terminates your app from task manager or force stop your app, it will start running again.

I'm looking for exactly this kind of functionality where my app remains responsible if terminated by the user.
I was checking all the distriqt ANEs, too, but did not find anything like this (@marchbold pls correct me if I overlooked it somehow..)

**Comment by itlancer** [comment]
Related feature requests:
https://github.com/airsdk/Adobe-Runtime-Support/issues/2468
https://github.com/airsdk/Adobe-Runtime-Support/issues/2818
And issue: https://tracker.adobe.com/#/view/AIR-4026124

**Reply by 2jfw** [reply]
Thanks for adding the relations 👍 (my researches for this had unfortunately not revealed them ..)

**Reply by marchbold** [reply]
@itlancer Gosh I had forgotten about that ios launching into the background issue. Gave up trying to get Adobe to acknowledge it.

**Comment by marchbold** [comment]
We have several services across our ANEs that perform certain functions in the background, we have created custom ones for clients as well. 

This is probably best kept native. Adding a custom service to an air app is pretty simple, and really just depends on the complexity of the functionality needed in the service.  

If you can describe what you are trying to do exactly I can see if we have something available.

**Reply by 2jfw** [reply]
Thanks Michael!
While in the background/terminated, the AIR app should be able to process specific code like reacting/listening to data from network like Sockets or NetGroups. The idea is to show something like a Native Notification when specific data is coming in which then opens the app or injects payload in the same manner as if the app was not terminated and the Notification pops up

**Reply by marchbold** [reply]
Have you considered using push notifications for this task?

**Reply by 2jfw** [reply]
Yes I did, but what speaks against this is that I would want to keep the scope of the app to local network and also keep it simple, too.
My understanding is, that push notifications would require google/apple services, also causing (additional) costs - I think I may need to revise the thought though in the case there is no other option ..

**Reply by marchbold** [reply]
Costs can depend on the scale but generally FCM and apple push are free I believe, the costs are generally associated with the server you'll need to maintain to send / manage the device registrations. There are web services you can use to handle this side of it (eg OneSignal) but they do introduce a cost.

Is this for a public app or for an installation (kiosk etc)?

Services on Android are still required to be "good citizens" so a service running continually with an active socket connection will likely result in a flag for battery usage etc. And since Oreo there have been a lot of additional requirements for background services https://developer.android.com/about/versions/oreo/background#services. 

You'd probably want to use a foreground service instead, which requires a visual presence (an ongoing notification is common, like what you'd see for a music player). But even so, I would suggest you consider using push notifications. This is definitely the most common approach for pushing data updates to devices.

**Reply by 2jfw** [reply]
Thanks a lot for the insight here - much appreciated!

The app would be public. Basically it's about data exchange between a PC and a phone and the phone being able to react on that (with a notification) in the case the app was terminated.

Interesting regarding the service running may causing battery issues and possibly being flagged - I understand the bigger picture behind this and reading a bit further I see the motives.

I will reconsider using push notifications then 👍
Thanks!
